server/src/main/java/org/opensearch/index/store/RemoteSegmentStoreDirectory.java

```diff
@@ -769,48 +836,60 @@ public void deleteStaleSegments(int lastNMetadataFilesToKeep) throws IOException
 
         Map<String, UploadedSegmentMetadata> activeSegmentFilesMetadataMap = new HashMap<>();
         Set<String> activeSegmentRemoteFilenames = new HashSet<>();
-        for (String metadataFile : sortedMetadataFileList) {
-            Map<String, UploadedSegmentMetadata> segmentMetadataMap = readMetadataFile(metadataFile).getMetadata();
-            activeSegmentFilesMetadataMap.putAll(segmentMetadataMap);
-            activeSegmentRemoteFilenames.addAll(
-                segmentMetadataMap.values().stream().map(metadata -> metadata.uploadedFilename).collect(Collectors.toSet())
-            );
-        }
+        Map<String, Boolean> metadataAlreadyProcessed = new HashMap<>();
+
+        Map<String, String> activePreviousMetadata = populatePreviousMetadata(
+            sortedMetadataFileList,
+            allLockFiles,
+            lastNMetadataFilesToKeep
+        );
+        Map<String, String> activeNextMetadata = populateNextMetadata(sortedMetadataFileList, allLockFiles, lastNMetadataFilesToKeep);
         Set<String> deletedSegmentFiles = new HashSet<>();
+        List<String> metadataFilesWithSuccessfulSegmentDeletion = new ArrayList<>();
         for (String metadataFile : metadataFilesToBeDeleted) {
             Map<String, UploadedSegmentMetadata> staleSegmentFilesMetadataMap = readMetadataFile(metadataFile).getMetadata();
             Set<String> staleSegmentRemoteFilenames = staleSegmentFilesMetadataMap.values()
                 .stream()
                 .map(metadata -> metadata.uploadedFilename)
                 .collect(Collectors.toSet());
+
+            processActiveSegments(
+                activePreviousMetadata.get(metadataFile),
+                activeSegmentFilesMetadataMap,
+                activeSegmentRemoteFilenames,
+                metadataAlreadyProcessed
+            );
+            processActiveSegments(
+                activeNextMetadata.get(metadataFile),
+                activeSegmentFilesMetadataMap,
+                activeSegmentRemoteFilenames,
+                metadataAlreadyProcessed
+            );
             AtomicBoolean deletionSuccessful = new AtomicBoolean(true);
+
+            List<String> staleSegmentFilesToBeDeleted = new ArrayList<>();
```

Comment:
With this approach , if some segment file deletion fails, we wouldn't delete it till the index is deleted .

Reply:
reverted this logic back to deleting one file at a time.